src/upchunk.ts

```diff
@@ -314,118 +494,154 @@ export class UpChunk  {
+        timeInterval: lastChunkInterval,
+        response: res,
+      });
+
+      this.attemptCount = 0;
+      this.chunkCount = (this.chunkCount ?? 0) + 1;
+      this.nextChunkRangeStart = this.nextChunkRangeStart + this.chunkByteSize;
+      if (this.dynamicChunkSize) {
+        let unevenChunkSize = this.chunkSize;
+        if (lastChunkInterval < 10) {
+          unevenChunkSize = Math.min(this.chunkSize * 2, this.maxChunkSize);
+        } else if (lastChunkInterval > 30) {
+          unevenChunkSize = Math.max(this.chunkSize / 2, this.minChunkSize);
+        }
+        // ensure it's a multiple of 256k
+        this.chunkSize = Math.ceil(unevenChunkSize / 256) * 256;
+
+        // Re-estimate the total number of chunks, by adding the completed
+        // chunks to the remaining chunks
+        const remainingChunks =
+          (this.file.size - this.nextChunkRangeStart) / this.chunkByteSize;
+        this.totalChunks = Math.ceil(this.chunkCount + remainingChunks);
+      }
+
+      return true;
+    };
+
+    // What to do if a chunk upload failed, potentially after retries
+    const failedChunkUploadCb = async (res: XhrResponse, _chunk?: Blob) => {
+      // Side effects
+      this.dispatch('error', {
+        message: `Server responded with ${
+          (res as XhrResponse).statusCode
+        }. Stopping upload.`,
+        chunk: this.chunkCount,
+        attempts: this.attemptCount,
+      });
+
+      return false;
+    };
+
+    // What to do if a chunk upload failed but is retriable and hasn't exceeded retry
+    // count
+    const retriableChunkUploadCb = async (
+      _res: XhrResponse | undefined,
+      _chunk?: Blob
+    ) => {
+      // Side effects
       this.dispatch('attemptFailure', {
         message: `An error occured uploading chunk ${this.chunkCount}. ${
           this.attempts - this.attemptCount
         } retries left.`,
         chunkNumber: this.chunkCount,
         attemptsLeft: this.attempts - this.attemptCount,
       });
-      return;
-    }
 
-    this.dispatch('error', {
-      message: `An error occured uploading chunk ${this.chunkCount}. No more retries, stopping upload`,
-      chunk: this.chunkCount,
-      attempts: this.attemptCount,
-    });
+      return new Promise<boolean>((resolve) => {
+        setTimeout(async () => {
+          // Handle mid-flight _paused/offline cases here by storing the
+          // "still retriable but yet to be uploaded chunk" in state.
+          // See also: `sendChunks()`
+          if (this._paused || this.offline) {
+            this.pendingChunk = chunk;
+            resolve(false);
+            return;
+          }
+          const chunkUploadSuccess = await this.sendChunkWithRetries(chunk);
+          resolve(chunkUploadSuccess);
+        }, this.delayBeforeAttempt * 1000);
+      });
+    };
+
+    let res: XhrResponse | undefined;
+    try {
+      this.attemptCount = this.attemptCount + 1;
+      this.lastChunkStart = new Date();
```

Comment:
nit (non-blocking): using `Date.now()` is a little more direct

Reply:
This was another one that was me trying to "leave code the same as much as possible". Since this is how it was previously implemented, I was planning on keeping as is.